rotkehlchen/chain/evm/decoding/curve_lend/balances.py

```diff
@@ -0,0 +1,164 @@
+            address=ZERO_ADDRESS,  # not used here
+            abi=CURVE_VAULT_CONTROLLER_ABI,
+            deployed_block=0,  # not used here
+        )
+        controller_list = []
+        user_address_list = []
+        for controller_address, user_addresses in controllers_with_balances.items():
+            for user_address in user_addresses:
+                controller_list.append(controller_address)
+                user_address_list.append(user_address)
+                calls.append((
+                    controller_address,
+                    token_contract.encode(
+                        method_name='user_state',
+                        arguments=[user_address],
+                    ),
+                ))
+
+        try:
+            call_output = self.evm_inquirer.multicall(calls=calls)
+        except RemoteError as e:
+            log.error(f'Failed to query Curve lending liabilities due to {e!s}')
+            return balances
+
+        for idx, result in enumerate(call_output):
+            user_state_data = token_contract.decode(
+                result=result,
+                method_name='user_state',
+                arguments=[user_address := user_address_list[idx]],
+            )[0]
+            collateral_amount = user_state_data[0]
+            borrowable_collateral_amount = user_state_data[1]
+            debt_amount = user_state_data[2]
+            if collateral_amount == 0 and borrowable_collateral_amount == 0 and debt_amount == 0:
+                continue  # user no longer has balances on this vault
+
+            # Get collateral and borrowed tokens for this vault
+            controller_address = controller_list[idx]
+            try:
+                call_output = self.evm_inquirer.multicall(calls=[
+                    (controller_address, token_contract.encode(method_name='collateral_token')),
+                    (controller_address, token_contract.encode(method_name='borrowed_token')),
+                ])
+            except RemoteError as e:
+                log.error(f'Failed to query tokens for Curve lending controller {controller_address} due to {e!s}')  # noqa: E501
+                return balances
+
+            collateral_token = EvmToken(evm_address_to_identifier(
+                address=token_contract.decode(
+                    result=call_output[0],
+                    method_name='collateral_token',
+                )[0],
+                chain_id=self.evm_inquirer.chain_id,
+                token_type=EvmTokenKind.ERC20,
+            ))
+            borrowed_token = EvmToken(evm_address_to_identifier(
+                address=token_contract.decode(
+                    result=call_output[1],
+                    method_name='borrowed_token',
+                )[0],
+                chain_id=self.evm_inquirer.chain_id,
+                token_type=EvmTokenKind.ERC20,
+            ))
+
+            # Add collateral to assets balances
+            for token, amount in [
```

Comment:
A generic question here. How common is it to have the same tokens here? As your current approach will do the price queries again and again for each token if they are part of different vault or different address balances.

A better approach would be to split the price queries off and query them later when you know the individual tokens.

Depends also on how they are structured and if they can repeat.

Reply:
Yes, actually with this approach the borrowed_token was getting its price queried twice already - and if the user has balances in multiple vaults, and/or mutiple addresses with balances in the same vaults the same tokens will come through this logic multiple times.

I've modified the logic here so that it should only query prices once for each token.